store/models/orm.go

```diff
@@ -160,6 +160,28 @@ func (orm *ORM) JobRunsWithStatus(statuses ...RunStatus) ([]JobRun, error) {
 	return runs, err
 }
 
+// AnyJobWithType returns true if there is at least one job associated with
+// the type name specified and false otherwise
+func (orm *ORM) AnyJobWithType(taskTypeName string) (bool, error) {
+	jobs := []JobSpec{}
+	err := orm.All(&jobs)
+	if err != nil {
+		return false, err
+	}
+	ts, err := NewTaskType(taskTypeName)
+	if err != nil {
+		return false, err
+	}
+	for i := range jobs {
+		for j := range jobs[i].Tasks {
+			if jobs[i].Tasks[j].Type == ts {
+				return true, nil
+			}
+		}
```

Comment:
I realize this is quite supoptimal, but storm does not support queries on arrays...

Reply:
Indeed. Let's try caching the job id in an array on bridge type when we write. Essentially denormalizing job id onto it's bridge type for faster retrievals:

```golang
// BridgeType is used for external adapters and has fields for
// the name of the adapter and its URL.
type BridgeType struct {
	Name                 TaskType `json:"name" storm:"id,unique"`
	URL                  WebURL   `json:"url"`
	DefaultConfirmations uint64   `json:"defaultConfirmations"`
	JobSpecIDs           []string `json:"jobSpecIds"` // new field to denormalize job ids
}

// later on in orm.go we update bridge types on write
func (orm *ORM) saveJob(....) {
  bridgeTasks := bridgeTypeTasksFor(savingJob)
  for bt in bridgeTasks {
    bt.JobSpecIDs = append(bt.JobSpecIDs, savingJob.ID)
    tx.Save(&bt)
    ... // continue to save job as originally planned
    tx.Commit()
  }
}
```

@se3000 can you confirm that this is the approach we would like to take? We add a field to `BridgeType` which will change what's returned by the API.

Reply:
@dimroc that could introduce a race condition if 2 jobs are created/updated at the roughly the same time. I feel like the logic to avoid race conditions could get pretty complicated.

Reply:
That would be ideal, but it's an optimization. Seems like this will work for now without having to add additional columns.

Reply:
@rupurt We save jobs inside a transaction to prevent race conditions like the one you mention:

```golang
func (orm *ORM) SaveJob(job *JobSpec) error {
	tx, err := orm.Begin(true) // begin tx
	if err != nil {
		return fmt.Errorf("error starting transaction: %+v", err)
	}
	defer tx.Rollback()

	for i := range job.Initiators {
		job.Initiators[i].JobID = job.ID
		if err := tx.Save(&job.Initiators[i]); err != nil {
			return fmt.Errorf("error saving Job Initiators: %+v", err)
		}
	}
	if err := tx.Save(job); err != nil {
		return fmt.Errorf("error saving job: %+v", err)
	}
	return tx.Commit()
}
```

Take note of the `orm.Begin(true)`. Does that alleviate your concerns?

Reply:
@dimroc I may be getting confused by `bridgeTasks := bridgeTypeTasksFor(savingJob)` 😄 

I was thinking this will return a slice of unique `BridgeType`'s.  If we do something like:

```go
func (orm *ORM) SaveJob(job *JobSpec) error {
	tx, err := orm.Begin(true) // begin tx
	if err != nil {
		return fmt.Errorf("error starting transaction: %+v", err)
	}
	defer tx.Rollback()

        bridgeTypes := bridgeTypesFor(savingJob)
        for bt in bridgeTypes {
          bt.JobSpecIDs = append(bt.JobSpecIDs, savingJob.ID)
          tx.Save(&bt)
        }

        ... // continue to save job as originally planned

        return tx.Commit()
}
```

If we create 2 `JobSpec` requests at the same time and `bridgeTypes := bridgeTypesFor(savingJob)` is assigned before either of the transactions are committed. When we go to append the `JobSpec` id's onto the `BridgeType`, they will be missing the `JobSpec` id from the other request. The last transaction to commit will win. Which will clobber the `append` operation from the other request.